process/block/preprocess/gasTracker.go

```diff
@@ -11,11 +13,82 @@ import (
 const noOverestimationFactor = uint64(100)
 
 type gasTracker struct {
-	shardCoordinator    sharding.Coordinator
-	economicsFee        process.FeeHandler
-	gasHandler          process.GasHandler
-	enableEpochsHandler common.EnableEpochsHandler
-	enableRoundsHandler common.EnableRoundsHandler
+	shardCoordinator     sharding.Coordinator
+	economicsFee         process.FeeHandler
+	gasHandler           process.GasHandler
+	enableEpochsHandler  common.EnableEpochsHandler
+	enableRoundsHandler  common.EnableRoundsHandler
+	mut                  sync.RWMutex
+	overEstimationFactor uint64
+	epochForLimits       uint32
+	roundForLimits       uint64
+}
+
+func newGasTracker(
+	shardCoordinator sharding.Coordinator,
+	gasHandler process.GasHandler,
+	economicsFee process.FeeHandler,
+	enableEpochsHandler common.EnableEpochsHandler,
+	enableRoundsHandler common.EnableRoundsHandler,
+) gasTracker {
+	return gasTracker{
+		shardCoordinator:     shardCoordinator,
+		economicsFee:         economicsFee,
+		gasHandler:           gasHandler,
+		enableEpochsHandler:  enableEpochsHandler,
+		enableRoundsHandler:  enableRoundsHandler,
+		overEstimationFactor: noOverestimationFactor,
+	}
+}
+
+// EpochConfirmed is called whenever a new epoch is confirmed
+func (gt *gasTracker) EpochConfirmed(epoch uint32, _ uint64) {
+	gt.mut.Lock()
+	defer gt.mut.Unlock()
+
+	gt.epochForLimits = epoch
+
+	// if already computed, only store the new epoch
+	if gt.overEstimationFactor != noOverestimationFactor {
+		return
+	}
+
+	isEpochFlagEnabled := gt.enableEpochsHandler.IsFlagEnabledInEpoch(common.SupernovaFlag, gt.epochForLimits)
+	if !isEpochFlagEnabled {
+		return
+	}
+
+	isRoundFlagEnabled := gt.enableRoundsHandler.IsFlagEnabledInRound(common.SupernovaRoundFlag, gt.roundForLimits)
+	if !isRoundFlagEnabled && epoch > 0 {
+		gt.epochForLimits = epoch - 1 // use the previous epoch until activation round
+	}
+}
+
+// RoundConfirmed is called whenever a new round is confirmed
+func (gt *gasTracker) RoundConfirmed(round uint64, _ uint64) {
+	gt.mut.Lock()
+	defer gt.mut.Unlock()
+
+	gt.roundForLimits = round
+
+	// if already computed, only store the new round
+	if gt.overEstimationFactor != noOverestimationFactor {
+		return
+	}
+
+	isRoundFlagEnabled := gt.enableRoundsHandler.IsFlagEnabledInRound(common.SupernovaRoundFlag, gt.roundForLimits)
+	if !isRoundFlagEnabled {
+		return
+	}
+
+	// new limits and overestimation should be enabled once the Supernova round is active
+	gt.overEstimationFactor = gt.economicsFee.BlockCapacityOverestimationFactor()
+	gt.epochForLimits = gt.epochForLimits + 1 // we save prev epoch until round is enabled too
```

Comment:
This comment was hard to understand at first. I'd say maybe we can have something a little more detailed like
```
// epoch was previously held at (currentEpoch - 1) until the Supernova round activated.
// now that the round is active, we advance epochForLimits to the real epoch.
```